### ISSUE TITLE
Unhandled Promise Rejection Can Crash App

### ISSUE THREAD

**Issue by GCAndrew** [segmentio/analytics-node#326]
As described in #320 , ‘flush()’ can cause an uncaught exception. Listening for uncaught exception events isn’t an ideal error handling pattern. 

This exception should be able to be handled at the call-site of ‘track()’ and similar functions.

**Comment by GCAndrew** [comment]
Example PR: https://github.com/segmentio/analytics-node/pull/327

**Comment by emont01** [comment]
@GCAndrew your PR adds a property that could be used to "turn off" error throwing once the axios request (promise) is rejected, I got nothing against this idea.

Reading the code, I am getting the impression that the idea here is to pass the possible error to any callback sent to flush and the individual message's callbacks, then throwing the exception again allows the promise to continue its normal course, which seems to be the problem here; my only question is: wouldn't it be better just to add an optional parameter to flush instead of making this a concern of the analytics client.

```js
  flush (callback, noUnhandledRejection = false) {
    callback = callback || noop
    ....
    .catch(err => {
        if (err.response) {
          const error = new Error(err.response.statusText)
          done(error)
          if (!noUnhandledRejection) {
            throw error
          }
        }

        done(err)
        if (!noUnhandledRejection) {
          throw err
        }
    })
```

**Comment by emont01** [comment]
@GCAndrew sounds good to me! I would suggest your PR makes the behavior the default 
```js
this.noUnhandledRejection = options.noUnhandledRejection || true
```

**Comment by GCAndrew** [comment]
@emont01 Great! Update my PR: https://github.com/segmentio/analytics-node/pull/327/commits/d6b7810b048153600d973a7025dfd86c222a460b

**Comment by emont01** [comment]
@GCAndrew yes!  The default value of noUnhandledRejection should be false IMHO, my main concerns right now is how setting it to true will break the retry behavior, we should let axios-retry do its thing before  handling any rejection.

To be honest I completely forgot about 'axios-retry' being used and after going over how that works [1] now I think some changes maybe necessary in either the tests, your PR or both.

axios-retry documentation does not mention any way to capture the "final error",  at that moment we could handle the exception and avoid an unhandled rejection, possibly something like [2].

I will try with my suggested try-catch changes and let you know if they work, let me know if you see a better way.

[1]
1. axios-retry catches any error and then checks if it is possible to retry
2. axios-retry will then return a promise with a delay or reject

[2]
``` js
  flush (callback) {
    ....
    try {
      return await this.axiosInstance.post(`${this.host}${this.path}`, data, req)
        .then(() => {
          done()
          return Promise.resolve(data)
        })
        .catch(err => {
          if (err.response) {
            const error = new Error(err.response.statusText)
            done(error)
            throw error
          }

          done(err)
          throw err
        })
    } catch (err) {
        if (!this.noUnhandledRejection) {
          throw err
        }
    }
  }
```

**Comment by tam-carre** [comment]
@nd4p90x Forgot to mention you, sorry!
